client/client-lib/src/mint/backup.rs

```diff
@@ -143,6 +157,118 @@ impl MintClient {
             .await?;
         Ok(())
     }
+
+    /// Fetch epochs in a given range and send them over `sender`
+    ///
+    /// Since WASM's `spawn` does not support join handles, we indicate
+    /// errors via `sender` itself.
+    ///
+    /// TODO: could be internal to recovery_loop?
+    async fn fetch_epochs(
+        &self,
+        epoch_range: RangeInclusive<u64>,
+        sender: mpsc::Sender<api::Result<EpochHistory>>,
+        task_handle: &TaskHandle,
+    ) {
+        // how to?
+        // let pubkey = self.config.epoch_pk_set.public_key();
+        let pubkey;
+
+        for epoch in epoch_range {
+            if task_handle.is_shutting_down() {
+                break;
+            }
+
+            match self.context.api.fetch_epoch_history(epoch, pubkey).await {
+                Ok(epoch_history) => {
+                    assert_eq!(epoch_history.outcome.epoch, epoch);
+                    // If the other side disconnected (probably due to an error),
+                    // we don't need to keep trying
+                    if sender.send(Ok(epoch_history)).await.is_err() {
+                        break;
+                    }
+                }
+                Err(e) => {
+                    if sender.send(Err(e)).await.is_err() {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn recovery_loop(
+        &self,
+        task_group: &mut TaskGroup,
+        backup: PlaintextEcashBackup,
+        treshold: usize,
+    ) -> Cancellable<Result<()>> {
+        let end_epoch = match self.context.api.fetch_consensus_block_height().await {
+            Ok(v) => v,
+            Err(e) => {
+                return Ok(Err(e.into()));
+            }
+        };
+        let epoch_range = backup.epoch..=backup.epoch;
+
+        // Since fetching epochs will be slow, we start a dedicated task to do it
+        let (tx, mut rx) = mpsc::channel(10);
+        let self_clone = self.clone();
+        task_group
+            .spawn("fetch epochs", {
+                let epoch_range = epoch_range.clone();
+                |task_handle| async move {
+                    self_clone
+                        .fetch_epochs(epoch_range, tx, &task_handle)
+                        .await
+                }
+            })
+            .await;
+
+        let mut nonce_tracker =
+            EcashRecoveryNonceTracker::from_backup(backup, self.secret.clone(), treshold, 1000);
+
+        for epoch in epoch_range {
+            // if `recv` returned `None` that means fetch_epoch finished prematurelly,
+            // withouth sending an `Err` which is supposed to mean `is_shutting_down() == true`
+            let epoch_history = match rx.recv().await {
+                Some(Ok(o)) => o,
+                Some(Err(e)) => return Ok(Err(e.into())),
+                None => return Err(Cancelled),
+            };
```

Comment:
It feels weird having fetching and processing in separate threads. If fetching is the bottle neck idk if that helps a lot?

Reply:
> It feels weird having fetching and processing in separate threads. If fetching is the bottle neck idk if that helps a lot?

I do parallel processing like this all the time, anytime there's higher volume/latency IO involved, either with other IO or anything CPU-intensive. It tends to give between 1x to 2x speedup, depending on how similar both sides tend to take.

I even have a library to do it nicely in forms of iterators: https://github.com/dpc/pariter . Too bad I don't have it for `async` code, then I could use Streams and it all would be quite elegant.